lib/engine/game/base.rb

```diff
@@ -192,11 +193,44 @@ def active_players
         @round.active_entities.map(&:owner)
       end
 
+      def process_undos(actions)
```

Comment:
i haven't tested this orr thought it through, but this was my idea


`arr = Array.new(actions.size)
undos = []

actions.each_with_index do |action, index|
  case action['type']
  when 'undo'
    i = arr.rindex { |a| a['type'] != 'message'}
    undos << [stack.delete(i), i]
  when 'redo'
    a, i = undos.pop
    arr[i] = a
  else
    undos.clear
    arr[index] = action
  end
end

arr.compact
`